### PR TITLE
Add an HOWTO for releasing Gitea

### PR DESCRIPTION
**Pull request by strk**
In the hope it'll be useful for as a policy for the future

### CODE REVIEW

**Comment by sapk** [comment]
Why not put that in CONTRIBUTING.md ?

**Comment by sapk** [comment]
and add part about : https://github.com/go-gitea/changelog ?

**Comment by strk** [comment]
Release is more for owners than for contributors.
I don't know about go-gitea/changelog, can you provide a PR against
this proposed branch if you know how that's supposed to be used ?

(maybe that one is more for CONTRIBUTING, if it implies contributors
providing their own changelog entry)

File: HOWTO_RELEASE.md
Comment (bkcsoft) [inline HOWTO_RELEASE.md]: there's a rouge ` in there :)

File: HOWTO_RELEASE.md
Comment (bkcsoft) [inline HOWTO_RELEASE.md]: use `` quotes

File: HOWTO_RELEASE.md
Comment (bkcsoft) [inline HOWTO_RELEASE.md]: code quote

**Review by bkcsoft — CHANGES_REQUESTED** [review]

**Comment by strk** [comment]
> there's a rouge ` in there :)

Fixed by a8e3ad98b38585dad60bc99b724580b6a50d2b68

> use `` quotes

Done with 646bf7fc

> code quote

Done with 646bf7fc

**Comment by lunny** [comment]
Contributors could know the process of release but don't perform that.

**Review by bkcsoft — APPROVED** [review]

**Comment by bkcsoft** [comment]
LGTM

@lunny I think it's a good idea to document the procedure, for future "release managers" (Owners currently) 🙂

**Comment by bkcsoft** [comment]
However in the future I'd like to just run `scripts/release v1.4.0` and be done with it :trollface:

**Comment by lunny** [comment]
@bkcsoft if we maintain PRs' labels correctly, we have done that via https://github.com/go-gitea/changelog .

File: HOWTO_RELEASE.md
Comment (bkcsoft) [inline HOWTO_RELEASE.md]: This should be after the branching section

File: HOWTO_RELEASE.md
Comment (bkcsoft) [inline HOWTO_RELEASE.md:18]: Please add a line below this that Drone automatically creates the Release and that it shouldn't be created manually (it doesn't break if one does, but just to document that it's not necessary)

Comment (strk) [reply HOWTO_RELEASE.md:18]: can you push directly to the branch associated to this PR ?

File: HOWTO_RELEASE.md
Comment (strk) [reply HOWTO_RELEASE.md]: It would make no difference, no  ? I like to idea of tagging *before* branching so that `git-describe` counts from latest release...

Comment (bkcsoft) [reply HOWTO_RELEASE.md]: Humm, that kinda makes sense 🤔

**Comment by bkcsoft** [comment]
LGTM

File: HOWTO_RELEASE.md
Comment (lunny) [inline HOWTO_RELEASE.md:4]: This document is in fact many different from the actual release process.

**Review by lunny — CHANGES_REQUESTED** [review]

**Comment by strk** [comment]
Can you document the actual release process so that it can be
understood by all maintainers, and amendments proposed ?

File: HOWTO_RELEASE.md
Comment (techknowlogick) [inline HOWTO_RELEASE.md:4]: Can you add a header to this document? Currently Contributing.md, the Readme, and others all have headers. Perhaps even adding `# Releasing Gitea` followed by two line breaks to the top of the file (or even "Release Procedure").

**Comment by strk** [comment]
> Can you add a header to this document?

See howto-release 1f194fc2

**Comment by strk** [comment]
@lunny your blocking this PR is irrational. You say that the procedure is wrong but do not provide the correct procedure. It doesn't help getting things right.

**Comment by lunny** [comment]
Please add how to release to `CONTRIBUTING.md` so that we don't need two document files. As I know, the simple releasing process is:

- Let $vmaj, $vmin and $vpat be Major, Minor and Patch version numbers, $vpat should be rc1, rc2, 0, 1, ...... $vmaj.$vmin will be kept the same as milestones on github or gitea in future.
- Before releasing, confirm all the version's milestone issues or PRs has been resolved. Then discuss the release on discord channel #maintainers and get agreed with almost all the owners and mergers. Or you can declare the version and if nobody against in about serval hours.
- If this is a big version, then you have to create a new branch named `release/v$vmaj.$vmin`
- Create PR for changelog on branch `release/v$vmaj.$vmin`
- After your PR reviewed and merged,  and CI passed, add a tag as `git tag -s -F release.notes v$vmaj.$vmin.$`, release.notes file could be a temporory file to only include the changelog this version which you added to `CHANGELOG.md`. 
- And then push the tag as `git push origin v$vmaj.$vmin.$`. Drone CI will automatically created a release and upload all the compiled binary. (But currently it didn't add the release notes automatically. Maybe we should fix that.)
- Send PR for changelog on branch `master`.
- Send PR to https://github.com/go-gitea/blog announcing the release.

**Comment by strk** [comment]
> - If this is a big version, then you have to create a new branch named `release/v$vmaj.$vmin`
> - Create PR for changelog on branch `release/v$vmaj.$vmin`

Why not doing the changelog in master, before the branch ?
That section of changelog will be good to have in future releases
anyway ?

> - After your PR reviewed and merged,  and CI passed, add a tag as `git tag -s -F release.notes v$vmaj.$vmin.$`, release.notes file could be a temporory file to only include the changelog this version which you added to `CHANGELOG.md`. 

As above, why not doing this in master branch ?

> - Send PR for changelog on branch `master`.

This PR would not be needed if we did it in master in the first place

**Comment by lunny** [comment]
@strk send it to master then send a back port to branch release/vvmaj.$vmin? There is no different. The changelog have to both in master and branch release.

**Comment by strk** [comment]
Send it to master and nowhere else. Release will come out from master branch. When release is finalized, and tagged, only then a branch is made and pushed, at that exact point branch == master == tag. From that point on things will start diverging again (with branch targetting next micro-release and master targetting next minor-release).

**Comment by lafriks** [comment]
@strk but that will make only RC1 tag being on master, is there really point in that?

**Comment by strk** [comment]
> @strk but that will make only RC1 tag being on master, is there really point in that?

That's a good question, RC releases management should be
part of this document, do they need to come out from branch
rather than from master ?

In any case, x.y.0 will be the only tag in master, even if
RC come out from master, with x.y.1 and subsequent still only
in the branch and not reachable from master. For git-describe
sake it may be a good idea to tag something like "x.y-start"
to indicate the start point of x.y.0 release procedure.

**Comment by lafriks** [comment]
@strk current workflow for us is following:
1) Create 1.x.0-RC1 tag
2) In master start working already on 1.(x+1).0 version
3) Wait for 2-3 weeks for 1.x branch to stabilize by backporting bugfixes reported in RC versions
4) Release 1.x.0 version

If you propose to tag everything till 1.x.0 version in master that would block new feature development for about a month and I don't think that is a good idea...

**Comment by techknowlogick** [comment]
I know this PR has stalled a bit, but some feedback on it is that using @lunny's comment from about I was able to successfully release 1.4.3 and 1.5.0.

I know there is some discussion on perhaps changing the process, however documenting it as it exists right now would help discussions about changing it in the future. I personally share same opinion of @lafriks from his last comment.

**Comment by lunny** [comment]
Maybe someone could create a new PR to instead this one. But I think I really don't want to read one more document.

**Comment by techknowlogick** [comment]
Closing due to #5054 replacement PR

**Pull request closed without merging**
